### PR TITLE
Add suspend() and resume() methods to DmDevice

### PR DESCRIPTION
**Pull request by mulkieran**
Rewrite methods that set tables to use these methods in what seems like the appropriate manner.

### CODE REVIEW

**Comment by mulkieran** [comment]
@agrover @trgill review, please? The questions are:
* Does it make sense at all to have resume and suspend apply to all devices?
* Is the order correct?

**Comment by mulkieran** [comment]
Blocks: https://github.com/stratis-storage/stratisd/pull/730.

File: src/thindev.rs [hunk 1]
```diff
@@ -143,10 +143,22 @@ impl DmDevice<ThinDevTargetTable> for ThinDev {
...
     fn size(&self) -> Sectors {
         self.table.table.length
     }
 
+    // ThinDevs can not be resumed or suspended directly.
+    // Therefore, this method always succeeds.
+    fn suspend(&mut self, _dm: &DM) -> DmResult<()> {
+        Ok(())
+    }
+
```
Comment (trgill) [inline src/thindev.rs]: can't we suspend it with?:

        let source_id = DevId::Name(self.name());
        dm.device_suspend(&source_id, DmFlags::DM_SUSPEND)?;

Comment (mulkieran) [reply src/thindev.rs]: Well, that would make the code more consistent...

**Comment by mulkieran** [comment]
umm, tests run forever, somehtings wrong.

**Comment by mulkieran** [comment]
I think I really have screwed up Jenkins this time...

**Comment by mulkieran** [comment]
Travis tests succeed, though.

**Comment by mulkieran** [comment]
Tests passing on Travis.

**Comment by trgill** [comment]
The device-mapper-test-sute has examples of pause/resume in the tests.  The commit_failure_sets_needs_check() test reloads the meta table and it suspends:

1. thin dev
2. thin pool
3. meta

The example makes me wonder if this type of pause should be done for meta/data expansion.

**Comment by mulkieran** [comment]
Was leaving bits of thinpool around, hopefully that's fixed.

**Comment by mulkieran** [comment]
rebased and updated.

**Comment by mulkieran** [comment]
Cache test hung for ever, but may have been due to this problem:

```
Feb 28 16:33:33 host-172-16-191-16 kernel: ---[ end trace 05a6c2288efb37ad ]---
Feb 28 16:33:33 host-172-16-191-16 kernel: RIP: __get_entry.part.8+0x5/0x10 [dm_cache_smq] RSP: ffffbda80ee2bb60
Feb 28 16:33:33 host-172-16-191-16 kernel: Code: 8b bb 40 01 00 00 e8 5b ba b2 c8 48 8b bb a8 00 00 00 e8 4f ba b2 c8 48 89 df 5b e9 f6 09 b5 c8 66 0f 1f 44 00 00 0f 1f 44 00 00 <0f> 0b 66 0f 1f 84 0
Feb 28 16:33:33 host-172-16-191-16 kernel: R13: 0000000000000001 R14: 00007f27329fe820 R15: 00007f27329fe6e0
Feb 28 16:33:33 host-172-16-191-16 kernel: R10: 00007f273242b378 R11: 0000000000000246 R12: 00007f27329fee01
Feb 28 16:33:33 host-172-16-191-16 kernel: RBP: 00007f27329fc910 R08: ffffffffffffffe0 R09: 0000000000000138
Feb 28 16:33:33 host-172-16-191-16 kernel: RDX: 00007f273242b240 RSI: 00000000c138fd06 RDI: 0000000000000003
Feb 28 16:33:33 host-172-16-191-16 kernel: RAX: ffffffffffffffda RBX: 00007f2732a20058 RCX: 00007f27333f08e7
Feb 28 16:33:33 host-172-16-191-16 kernel: RSP: 002b:00007f27329fbdf8 EFLAGS: 00000246 ORIG_RAX: 0000000000000010
Feb 28 16:33:33 host-172-16-191-16 kernel: RIP: 0033:0x7f27333f08e7
Feb 28 16:33:33 host-172-16-191-16 kernel:  entry_SYSCALL_64_after_hwframe+0x21/0x86
Feb 28 16:33:33 host-172-16-191-16 kernel:  do_syscall_64+0x75/0x180
Feb 28 16:33:33 host-172-16-191-16 kernel:  SyS_ioctl+0x74/0x80
Feb 28 16:33:33 host-172-16-191-16 kernel:  do_vfs_ioctl+0xa4/0x620
Feb 28 16:33:33 host-172-16-191-16 kernel:  dm_ctl_ioctl+0xa/0x10
Feb 28 16:33:33 host-172-16-191-16 kernel:  ctl_ioctl+0x1be/0x470
Feb 28 16:33:33 host-172-16-191-16 kernel:  dev_suspend+0x14d/0x2a0
Feb 28 16:33:33 host-172-16-191-16 kernel:  dm_suspend+0xb5/0xc0
Feb 28 16:33:33 host-172-16-191-16 kernel:  dm_table_postsuspend_targets+0x45/0x60
Feb 28 16:33:33 host-172-16-191-16 kernel:  ? table_load+0x340/0x340
Feb 28 16:33:33 host-172-16-191-16 kernel:  cache_postsuspend+0x2a7/0x320 [dm_cache]
Feb 28 16:33:33 host-172-16-191-16 kernel:  dm_cache_write_hints+0xd2/0x100 [dm_cache]
Feb 28 16:33:33 host-172-16-191-16 kernel:  dm_array_new+0x15a/0x1f0 [dm_persistent_data]
Feb 28 16:33:33 host-172-16-191-16 kernel:  get_hint+0x1c/0x30 [dm_cache]
Feb 28 16:33:33 host-172-16-191-16 kernel:  smq_get_hint+0x3c/0x40 [dm_cache_smq]
Feb 28 16:33:33 host-172-16-191-16 kernel: Call Trace:
Feb 28 16:33:33 host-172-16-191-16 kernel: DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
Feb 28 16:33:33 host-172-16-191-16 kernel: DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
Feb 28 16:33:33 host-172-16-191-16 kernel: CR2: 00007f2732445000 CR3: 00000003b041a003 CR4: 00000000003606f0
Feb 28 16:33:33 host-172-16-191-16 kernel: CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
Feb 28 16:33:33 host-172-16-191-16 kernel: FS:  00007f27329ff700(0000) GS:ffffa0a03fc00000(0000) knlGS:0000000000000000
Feb 28 16:33:33 host-172-16-191-16 kernel: R13: 0000000000000001 R14: ffffffffc06f0760 R15: 0000000000000002
Feb 28 16:33:33 host-172-16-191-16 kernel: R10: ffffbda80ee2bb18 R11: 0000000000000040 R12: 0000000000000000
Feb 28 16:33:33 host-172-16-191-16 kernel: RBP: ffffa09fd3cd78a8 R08: ffffa0a03fc25080 R09: ffffa0a0176a6000
Feb 28 16:33:33 host-172-16-191-16 kernel: RDX: ffffbda80eeec818 RSI: 0000000000000501 RDI: ffffa09fd4c2f000
Feb 28 16:33:33 host-172-16-191-16 kernel: RAX: ffffa09fd4c2f0a8 RBX: ffffa0a0176a601c RCX: 0000000000000f03
Feb 28 16:33:33 host-172-16-191-16 kernel: RSP: 0018:ffffbda80ee2bb60 EFLAGS: 00010246
Feb 28 16:33:33 host-172-16-191-16 kernel: RIP: 0010:__get_entry.part.8+0x5/0x10 [dm_cache_smq]
Feb 28 16:33:33 host-172-16-191-16 kernel: Hardware name: Red Hat OpenStack Compute, BIOS 1.10.2-3.el7_4.1 04/01/2014
Feb 28 16:33:33 host-172-16-191-16 kernel: CPU: 0 PID: 27670 Comm: cachedev::tests Not tainted 4.15.4-300.fc27.x86_64 #1
Feb 28 16:33:33 host-172-16-191-16 kernel: Modules linked in: dm_thin_pool dm_cache_smq dm_cache dm_persistent_data dm_bio_prison loop binfmt_misc ip6t_rpfilter ip6t_REJECT nf_reject_ipv6 xt_conntrac
Feb 28 16:33:33 host-172-16-191-16 kernel: invalid opcode: 0000 [#1] SMP PTI
Feb 28 16:33:33 host-172-16-191-16 kernel: kernel BUG at drivers/md/dm-cache-policy-smq.c:90!
Feb 28 16:33:33 host-172-16-191-16 kernel: ------------[ cut here ]------------
```

**Comment by trgill** [comment]
Yikes.  I think the test nodes are fairly up to date, but I'm happy to update to the latest if you'd like.

**Comment by mulkieran** [comment]
@trgill We got exactly the same trace again. If updating makes this problem go away, I'm all for it!

**Comment by mulkieran** [comment]
> The example makes me wonder if this type of pause should be done for meta/data expansion.

@trgill I think you must be right about this. I'ld like to get the cache devices cooperating a little better before I get started on the thinpool stuff, though.

**Comment by mulkieran** [comment]
Well, looking at the state of the various devices, we can conclude the following.

1. The cache device is suspended.
2. The cache sub-device is suspended.
3. The new table for the cache sub-device is succesfully loaded.
4. The cache sub-device is resumed.
5. The cache device fails to resume properly. It stays suspended.

Here's the tables at that time:

```
[root@host-172-16-191-16 ~]# dmsetup table cache
0 2097152 cache 253:0 253:1 253:2 64 1 writethrough default 0
[root@host-172-16-191-16 ~]# dmsetup table cache-cache
0 64 linear 7:0 4096
64 64 linear 7:2 0
[root@host-172-16-191-16 ~]# dmsetup table cache-meta
0 4096 linear 7:0 0
[root@host-172-16-191-16 ~]# dmsetup table cache-origin
0 2097152 linear 7:1 0
```

**Comment by mulkieran** [comment]
The offending devicemapper code:

```
static struct entry *__get_entry(struct entry_space *es, unsigned block)
{
    struct entry *e;

    e = es->begin + block;
    BUG_ON(e >= es->end);

    return e;
}

```

**Comment by mulkieran** [comment]
Superceded by #283.

**Pull request closed without merging**